pkg/fleet-manager/application_controller.go

```diff
@@ -0,0 +1,384 @@
+	if err := a.Client.Get(ctx, fleetKey, fleet); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("fleet does not exist", "fleet", fleetKey)
+			return ctrl.Result{RequeueAfter: RequeueAfter}, nil
+		}
+
+		// Log error and requeue request if error occurred during fleet retrieval
+		log.Error(err, "failed to find fleet", "fleet", fleetKey)
+		return ctrl.Result{}, err
+	}
+
+	// label the fleet
+	if fleet.GetLabels() == nil {
+		fleet.SetLabels(make(map[string]string))
+	}
+	// todo: there is a bug here if more than one application refer the same fleet: the new one will override the old
+	//  but it is needed to reconcile application when the fleet' cluster is all joined
+	if fleet.GetLabels()[ApplicationLabel] != app.Name {
+		labels := fleet.GetLabels()
+		labels[ApplicationLabel] = app.Name
+		fleet.SetLabels(labels)
+		err := a.Update(ctx, fleet)
+		if err != nil {
+			log.Error(err, "unable to label fleet", "fleet", fleet.Name)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Handle deletion reconciliation loop.
+	if app.DeletionTimestamp != nil {
+		return a.reconcileDelete(ctx, app, fleet)
+	}
+
+	// Handle normal loop.
+	return a.reconcile(ctx, app, fleet)
+}
+
+func (a *ApplicationManager) reconcile(ctx context.Context, app *applicationapi.Application, fleet *fleetapi.Fleet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if result, err := a.reconcileInitializeParameters(ctx, app); err != nil || result.RequeueAfter > 0 {
+		log.Error(err, "failed to reconcileInitializeParameters")
+		return result, err
+	}
+
+	if result, err := a.reconcileSyncResources(ctx, app, fleet); err != nil || result.RequeueAfter > 0 {
+		log.Error(err, "failed to reconcileSyncResources")
+
+		return result, err
+	}
+
+	if result, err := a.reconcileSyncStatus(ctx, app); err != nil || result.RequeueAfter > 0 {
+		log.Error(err, "failed to reconcileSyncStatus")
+
+		return result, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcileInitializeParameters initializes the parameters for the given application. It also ensures some parameters is valid.
+func (a *ApplicationManager) reconcileInitializeParameters(ctx context.Context, app *applicationapi.Application) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Find the source kind for the given application. Return an error if no source or more than one source is specified.
+	sourceKind, err := findSourceKind(app)
+	if err != nil {
+		log.Error(err, "failed to find source kind")
+		return ctrl.Result{}, err
+	}
+
+	// Set the source kind for the application.
+	app.Spec.Source.Kind = sourceKind
+
+	// Iterate over all sync policies.
+	for index, policy := range app.Spec.SyncPolicy {
+		// If no policy name is specified, set a default in the format `<application name>-<index>`.
+		if len(policy.Name) == 0 {
+			policy.Name = app.Name + "-" + strconv.Itoa(index)
+		}
+
+		// Currently, only two pairs of source types are supported: 'gitRepo + kustomization' and 'helmRepo + helmRelease'.
```

Comment:
I can see in flux there is no such limitation.

Reply:
todo